osu.Server.Queues.ScoreStatisticsProcessor/Commands/Queue/WatchHighScoresCommand.cs

```diff
@@ -136,18 +140,44 @@ public async Task<int> OnExecuteAsync(CancellationToken cancellationToken)
                                              .OrderBy(s => s.score_id)
                                              .ToArray();
 
-                    var pendingProcessing = highScores.FirstOrDefault(s => s.status == 0);
+                    var lastPending = pendingProcessing;
+                    pendingProcessing = highScores.FirstOrDefault(s => s.status == 0);
+
+                    const int sleep = 500;
 
                     if (pendingProcessing != null)
                     {
-                        Console.WriteLine($"Waiting on processing of (score_id: {pendingProcessing.score_id} queue_id: {pendingProcessing.queue_id})");
-                        Thread.Sleep(500);
-                        continue;
+                        // the pending score may have changed, so reset the wait time at this point.
+                        if (pendingProcessing.score_id != lastPending?.score_id)
+                        {
+                            pendingProcessingWaitTime = sleep;
+
+                            Console.WriteLine($"Waiting on processing of (score_id: {pendingProcessing.score_id} queue_id: {pendingProcessing.queue_id})");
+                            Thread.Sleep(sleep);
+                            continue;
+                        }
+
+                        // 5 minute timeout
+                        if (pendingProcessingWaitTime > 300000)
+                        {
+                            Console.WriteLine($"Importing score which was not processed due to timeout (score_id: {pendingProcessing.score_id} queue_id: {pendingProcessing.queue_id})");
+                        }
```

Comment:
Isn't this going to import not only the score with id `pendingProcesing.score_id`, but the rest of the involved batch as well, which may contain other unprocessed scores? Should that be logged at least?

Reply:
Sounds fair to only process up to the currently pending item and continue afterwards for any furthers.